### ISSUE TITLE
Add shortcut to restart and restore to "flutter run"

### ISSUE THREAD

**Issue by goderbauer** [flutter/flutter#63233]
While `flutter run`ning an app in debug mode it would be nice if there were a shortcut to restore the application from its current restoration state. That would make debugging and experimenting with state restoration a lot easier.

Currently, you'll have to disable "keep activities" in the Android's developer settings and then, background your app, and bring it back to foreground to trigger a restore.

This is currently also not compatible with hot reload/restart. If you hot related changes into your app, those are all gone when the app comes back from background.

**Comment by jonahwilliams** [comment]
Curious, what doesn't work with hot restart?

**Comment by goderbauer** [comment]
Hot restart throws away the restoration data, which seems sensible as hot restart is supposed to lose all app state and re-initialize the app from scratch. 

We basically need a mode that is like hot restart, but applies the restoration data that was collected just before the hot restart.

**Comment by jonahwilliams** [comment]
Ahh, I'm curious why that is actually. The restoration data is saved by the embedder right? Or do we just need to flush the data before performing the restart

**Comment by goderbauer** [comment]
Because we wired it up to clear the restoration data in the embedder on restart: https://github.com/flutter/engine/blob/b77e45cbd883e7f34bdf952b3283f9de379378b6/shell/platform/android/io/flutter/embedding/engine/FlutterEngine.java#L109


... which to me makes sense because there's no guarantee that the restoration data collected with the old code can be applied to the code that was loaded in with hot restart.

**Comment by goderbauer** [comment]
Thinking about this more, the new shortcut should probably not load in new code. It should just restart the app with the existing code and apply the restoration data.

**Comment by jonahwilliams** [comment]
Yeah that makes sense in most cases. I think the difficulty with this feature is that the tool doesn't really understand how to restart applications, besides through the hot restart cycle. If it was feasible to make the state destruction conditional, that might be one possibility.
